liveblocks/liveblocks
# Example react native todo list upgrade

### Upgrading React Native Todo List example

The example with React Native and Liveblocks doesn't work with the latest environment (OS, Xcode, ...) #1499. 
This PR aims to upgrade the example's library `react-native@0.68.2` to `react-native@0.73.6` and make it work with the latest `liveblocks/client` package.

## React Native upgrade

The upgrade has been applied in this order:
- 0.68.2 --> 0.70.0
- 0.70.0 --> 0.73.0
- 0.73.0 --> 0.73.6 (latest stable)

## Liveblocks adjustments

After upgrading `liveblocks/client` two errors are displayed in RN engine: 
- `window.addEventListener is undefined`
- `window.postMessage is undefined`

The first error was handled in `liveblocks/client@0.17.0`. I am not sure if we can say it's a regression. I didn't check in which version it has stopped working.
It refers to handling reconnection if I am correct.  This problem was addressed here #261.

The second error is related to the DevTools feature. Investigation is needed to see if it's possible to work with the debugger mode of RN (a new debugger is currently in experimentation with the package 0.73.6). 

Early in the app's lifecycle, the following lines have been added (index.js) to fix the 2 errors:
```
window.addEventListener = () => { };
window.postMessage = () => { };
```

`URL` isn't available in RN. That's why a polyfill is needed. It didn't work with the package `core-js`. 
It worked with the package `react-native-url-polyfill`.
There is an attempt to integrate `URL` as part of RN but the PR has been rejected https://github.com/facebook/react-native/pull/38675.

Here is the code to make `URL` available (index.js): 
```
import { URL } from "react-native-url-polyfill";
global.URL = URL;
```

Steps to test:
- ensure the dev environment is installed and up to date https://reactnative.dev/docs/environment-setup
- delete `node_modules`
- delete `ios/Pods/`
- uninstall `react-native` if it's installed globally
- `npm install`
- `ios/pod install`
- `npm run ios`
- `npm run android`

## QA
Here the video of the Todo List example with RN: https://youtu.be/0K0ux6jluUk?si=hIDPPy67z5Ixq-GL
